lua/rocks/operations/lock.lua

```diff
@@ -0,0 +1,78 @@
+---@mod rocks.operations.lock
+--
+-- Copyright (C) 2024 Neorocks Org.
+--
+-- License:    GPLv3
+-- Created:    04 Apr 2024
+-- Updated:    04 Apr 2024
+-- Homepage:   https://github.com/nvim-neorocks/rocks.nvim
+-- Maintainers: NTBBloodbath <bloodbathalchemist@protonmail.com>, Vhyrro <vhyrro@gmail.com>, mrcjkb <marc@jakobi.dev>
+--
+---@brief [[
+--
+-- Lockfile management.
+--
+---@brief ]]
+
+local config = require("rocks.config.internal")
+local fs = require("rocks.fs")
+
+local lock = {}
+
+---@param reset boolean
+local function parse_rocks_lock(reset)
+    local lockfile = reset and "" or fs.read_or_create(config.lockfile_path, "")
+    return require("toml_edit").parse(lockfile)
+end
+
+---@param rock_name? rock_name
+function lock.update_lockfile(rock_name)
+    local luarocks_lockfiles = vim.iter(vim.api.nvim_get_runtime_file("luarocks.lock", true))
+        :filter(function(path)
+            return not rock_name or path:find(rock_name .. "/[^%/]+/luarocks.lock$") ~= nil
+        end)
+        :totable()
+    local reset = rock_name == nil
+    local rocks_lock = parse_rocks_lock(reset)
+    for _, luarocks_lockfile in ipairs(luarocks_lockfiles) do
+        local rock_key = rock_name or luarocks_lockfile:match("/([^%/]+)/[^%/]+/luarocks.lock$")
+        if rock_key then
+            local ok, loader = pcall(loadfile, luarocks_lockfile)
+            if not ok or not loader then
+                return
+            end
+            local success, luarocks_lock_tbl = pcall(loader)
+            if not success or not luarocks_lock_tbl or not luarocks_lock_tbl.dependencies then
+                return
+            end
+            rocks_lock[rock_key] = {}
+            for dep, version in pairs(luarocks_lock_tbl.dependencies) do
+                if dep ~= "lua" then
```

Comment:
I think we should also exclude non semver versions from the lockfile.

Pinning `scm` could lead to subtle bugs. Perhaps we should also create a luarocks issue.